Add aimlab extension

whitespacecode opened the pull request:
## Description

Track Player profiles, Seasonal Leaderboards and Top Creators from Aimlab

## Screencast
https://github.com/raycast/extensions/assets/37139936/7e0f40ff-5a56-4819-8f52-5fc1e36ef525

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

thomaslombart commented on extensions/aimlab/README.md:
```diff
@@ -0,0 +1,37 @@
+
+### **Leaderboards**
+
+In theory there should not be more then one active season. Aimlab will determine the currently active season.
```
```suggestion
In theory, there should not be more than one active season. Aimlab will determine the currently active season.
```

thomaslombart commented on extensions/aimlab/src/components/Creator.tsx:
```diff
@@ -0,0 +1,49 @@
+      keywords={[author.username]}
+      actions={
+        <ActionPanel title="Top Creator Actions">
+          <Action
+            title="Show Player Details"
+            icon={Icon.ArrowRight}
+            onAction={() => push(<UserProfile username={author.username} />)}
+          />
```
`push` is designed for programmatic navigation. You can use `Action.Push` instead.

thomaslombart commented on extensions/aimlab/src/components/Creator.tsx:
```diff
@@ -0,0 +1,49 @@
+          <Action.OpenInBrowser
+            title="View Player on Aimlab"
+            url={"https://aimlab.gg/u/" + encodeURIComponent(author.username)}
+            icon={Icon.Globe}
```
`Action.OpenInBrowser` has the `Icon.Globe` as an icon by default.

```suggestion
```

thomaslombart commented on extensions/aimlab/src/components/Season.tsx:
```diff
@@ -0,0 +1,33 @@
+        })
+      ) : (
+        <List.EmptyView
+          icon={{ source: "logo.png" }}
```
```suggestion
          icon="logo.png"
```

thomaslombart commented on extensions/aimlab/src/components/Season.tsx:
```diff
@@ -0,0 +1,33 @@
+          return <SeasonTaskComponent key={task.taskId} season={season} seasonTask={task} />;
+        })
+      ) : (
+        <List.EmptyView
+          icon={{ source: "logo.png" }}
+          title="No tasks found!"
+          description="Please visit the website instead"
+        />
```
You can add an `actions` prop to the `List.EmptyView` so that users can visit the website from the action panel.

thomaslombart commented on extensions/aimlab/src/hooks/usePlaysAgg.ts:
```diff
@@ -0,0 +1,29 @@
+      }
+
+      const playeragg: PlayerAgg[] = data.map(
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Is the `any` necessary? You could type the results from the API instead.

thomaslombart commented on extensions/aimlab/src/utils/queries.ts:
```diff
@@ -0,0 +1,261 @@
+  }
+`;
+
+// leaderboardInput => {
+//     "leaderboardInput":{
+//       "clientId": "aimlab",
+//       "limit":200,
+//       "offset":0,
+//       "taskId":"id",
+//       "taskMode": 0,
+//       "weaponId":"weapon"
+//   },
+//   "window":
+//   {
+//       "period":"month/week/year",
+//       "value":"yr-mo-day"
+//   }
+// }
```
Seems like a leftover.

thomaslombart commented on extensions/aimlab/src/views/profile.tsx:
```diff
@@ -0,0 +1,54 @@
+};
+
+const UserProfile = ({ username }: PropTypes) => {
+  const preferences = getPreferenceValues<Preferences>();
```
Note that you can use the types from `raycast-env.d.ts` to automatically get preferences types: https://developers.raycast.com/changelog#new-3.

thomaslombart reviewed (changes requested):
Congrats on the extension! I have some feedback on it.

## Flickering when searching for players

The `Aimlab Profile` command seems to flicker. Could you get rid of them? Ideally, the user should see the loading indicator, then the results and nothing in between.

https://github.com/raycast/extensions/assets/16003285/c3c1c506-b314-4765-b903-ed2105d54cc1

## Default icon for creators

I've noticed there are no icons for certain creators. Would it be possible to show a default icon for them to have a consistent result?

![CleanShot 2023-07-03 at 16 31 35@2x](https://github.com/raycast/extensions/assets/16003285/dfb4a3ff-3b36-45f1-abe9-b132bf36c578)

## Make searching profiles more flexible

Currently, you have to enter the exact username to get the result increasing the odds of landing on the empty results screen. Two things you can do:

- Ignore the case (lowercase/uppercase), for example, `vt matty` should return `VT Matty` profile.
- Ignore the whitespaces, for example, `vtmatty` should return `VT Matty` profile.

You can also see one profile at a time which makes it a bit weird given the components that are used. Why not displaying all profiles that match the search query?

<img width="846" alt="CleanShot 2023-07-03 at 16 43 24@2x" src="https://github.com/raycast/extensions/assets/16003285/4c5136ba-22f9-498b-a8e8-696fed15c598">

whitespacecode commented:
> Congrats on the extension! I have some feedback on it.
> 
> ## Flickering when searching for players
> The `Aimlab Profile` command seems to flicker. Could you get rid of them? Ideally, the user should see the loading indicator, then the results and nothing in between.
> 
>  CleanShot.2023-07-03.at.16.32.59.mp4 
> ## Default icon for creators
> I've noticed there are no icons for certain creators. Would it be possible to show a default icon for them to have a consistent result?
> 
> ![CleanShot 2023-07-03 at 16 31 35@2x](https://user-images.githubusercontent.com/16003285/250586431-dfb4a3ff-3b36-45f1-abe9-b132bf36c578.png)
> 
> ## Make searching profiles more flexible
> Currently, you have to enter the exact username to get the result increasing the odds of landing on the empty results screen. Two things you can do:
> 
> * Ignore the case (lowercase/uppercase), for example, `vt matty` should return `VT Matty` profile.
> * Ignore the whitespaces, for example, `vtmatty` should return `VT Matty` profile.
> 
> You can also see one profile at a time which makes it a bit weird given the components that are used. Why not displaying all profiles that match the search query?
> 
> <img alt="CleanShot 2023-07-03 at 16 43 24@2x" width="846" src="https://user-images.githubusercontent.com/16003285/250588160-4c5136ba-22f9-498b-a8e8-696fed15c598.png">

Fixed the flickering, added a default icon for creators and added a screencast.
One thing i want to clarify about the profile search. As i would love to show a list of players that is not possible. They only provide a exact search on username, there is no other query. So knowing the search on username should be unique what components should i use instead of a listview ?

thomaslombart commented:
Hey @whitespacecode, in this case, there's no other components you can use as the `List` component is the only one you can use to search for something.

whitespacecode commented:
Okay i see. I will look into the flickering. 
For the seasons once you fetch the season(s) it returns the season with tasks but the tasks only have an id. So i fetch the taskInformation separately. I will change it so when fetching the season i also fetch the taskInformation directly.

whitespacecode commented:
> I've tested it again but I can still see some flickering happening. 

I fetched all the data to show all the results at once. 
One thing i can't seem to simulate is the flickering. Do you have slow internet or are there other ways i could test this ?

https://github.com/raycast/extensions/assets/37139936/3096b05a-c94a-48c2-a03b-4801d0d36c9f

thomaslombart reviewed (approved):
🚀
